// PROTOBUF definitions
syntax = "proto3";

// a video frame
message VideoFrame {
  int32 number = 1;
  int64 timestamp = 2;
  bytes img = 3;
}

// a YoLo detection request
message YoloDetectionRequest {
  VideoFrame frame = 1;
  int32 max_results = 2;
}

// a face detection request
message FaceDetectionRequest {
  VideoFrame frame = 1;
  int32 max_results = 2;
}

// an object detection request
message ObjectDetectionRequest {
  int32 user_id = 1;
  VideoFrame frame = 2;
  int32 wantface = 3;
  int32 wantyolo = 4;
  int32 max_results = 5;
}

// a detected Yolo
message DetectedYolo {
  string name = 1;
  float confidence = 2;
  int32 x = 3;
  int32 y = 4;
  int32 w = 5;
  int32 h = 6;
  uint32 r = 7;
  uint32 g = 8;
  uint32 b = 9;
}

// a detected face
message DetectedFace {
  float confidence = 1;
  int32 x = 2;
  int32 y = 3;
  int32 w = 4;
  int32 h = 5;
  uint32 r = 6;
  uint32 g = 7;
  uint32 b = 8;
}

// a detected object
message DetectedObject {
  string name = 1;
  enum Eng {
    FACE = 0;
    YOLO = 1;
  };
  Eng eng = 2;
  float confidence = 3;
  int32 x = 4;
  int32 y = 5;
  int32 w = 6;
  int32 h = 7;
  uint32 r = 8;
  uint32 g = 9;
  uint32 b = 10;
}

message YoloDetectionResponse {
  int32 number = 1;
  int64 timestamp = 2;
  repeated DetectedYolo yolos = 3;
}

message FaceDetectionResponse {
  int32 number = 1;
  int64 timestamp = 2;
  repeated DetectedFace faces = 3;
}

message ObjectDetectionResponse {
  int32 number = 1;
  int64 timestamp = 2;
  repeated DetectedObject objects = 3;
  fixed32 ipv4 = 4;
  string name = 5;
}

// detect a yolo
service YoloDetection {
  rpc YoloDetect(YoloDetectionRequest) returns (YoloDetectionResponse);
}

// detect a face
service FaceDetection {
  rpc FaceDetect(FaceDetectionRequest) returns (FaceDetectionResponse);
}

// detect a generic object
service ObjectDetection {
  rpc ObjectDetect(ObjectDetectionRequest) returns (ObjectDetectionResponse);
}